File: app/src/main/java/com/aar/app/wordsearch/data/xml/WordThemeDataXmlLoader.java
package com.aar.app.wordsearch.data.xml;

import android.content.Context;
import android.content.res.AssetManager;

import com.aar.app.wordsearch.model.GameTheme;
import com.aar.app.wordsearch.model.Word;

import org.xml.sax.InputSource;
import org.xml.sax.XMLReader;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import javax.xml.parsers.SAXParserFactory;

public class WordThemeDataXmlLoader {

    private static final String ASSET_BASE_FOLDER = "words";

    private AssetManager mAssetManager;

    private List<Word> mWords;
    private List<GameTheme> mGameThemes;

    public WordThemeDataXmlLoader(Context context) {
        mAssetManager = context.getAssets();
    }

    public List<Word> getWords() {
        if (mWords == null) {
            loadData();
        }
        return mWords;
    }

    public List<GameTheme> getGameThemes() {
        if (mGameThemes == null) {
            loadData();
        }
        return mGameThemes;
    }

    public void release() {
        mWords.clear();
        mGameThemes.clear();
        mWords = null;
        mGameThemes = null;
    }

    private void loadData() {
        try {
            XMLReader reader = SAXParserFactory.newInstance().newSAXParser().getXMLReader();
            SaxWordThemeHandler handler = new SaxWordThemeHandler();
            reader.setContentHandler(handler);

            for (String fileName : getAssetFilePaths()) {
                reader.parse(getInputSource(fileName));
            }

            mWords = handler.getWords();
            mGameThemes = handler.getGameThemes();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private InputSource getInputSource(String fileName) throws IOException {
        return new InputSource(mAssetManager.open(fileName));
    }

    private List<String> getAssetFilePaths() {
        List<String> filePaths = new ArrayList<>();
        try {
            String fileNames[] = mAssetManager.list(ASSET_BASE_FOLDER);
            for (String fileName : fileNames) {
                filePaths.add(ASSET_BASE_FOLDER + "/" + fileName);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return filePaths;
    }
}

File: app/src/main/java/com/aar/app/wsp/data/xml/WordThemeDataXmlLoader.kt
package com.aar.app.wsp.data.xml

import android.content.Context
import android.content.res.AssetManager
import com.aar.app.wsp.model.GameTheme
import com.aar.app.wsp.model.Word
import org.xml.sax.InputSource
import java.io.IOException
import java.util.*
import javax.xml.parsers.SAXParserFactory

class WordThemeDataXmlLoader(context: Context) {
    private val assetManager: AssetManager = context.assets
    private var _words: List<Word>? = null
    private var _gameThemes: List<GameTheme>? = null

    val words: List<Word>
        get() {
            if (_words == null) {
                loadData()
            }
            return _words ?: emptyList()
        }

    val gameThemes: List<GameTheme>
        get() {
            if (_gameThemes == null) {
                loadData()
            }
            return _gameThemes ?: emptyList()
        }

    fun release() {
        _words = null
        _gameThemes = null
    }

    private fun loadData() {
        try {
            val reader = SAXParserFactory.newInstance().newSAXParser().xmlReader
            val handler = SaxWordThemeHandler()
            reader.contentHandler = handler
            for (fileName in assetFilePaths) {
                reader.parse(getInputSource(fileName))
            }
            _words = handler.words
            _gameThemes = handler.gameThemes
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    @Throws(IOException::class)
    private fun getInputSource(fileName: String): InputSource {
        return InputSource(assetManager.open(fileName))
    }

    private val assetFilePaths: List<String>
        get() {
            val filePaths: MutableList<String> = ArrayList()
            try {
                val fileNames = assetManager.list(ASSET_BASE_FOLDER)
                for (fileName in fileNames!!) {
                    filePaths.add("$ASSET_BASE_FOLDER/$fileName")
                }
            } catch (e: IOException) {
                e.printStackTrace()
            }
            return filePaths
        }

    companion object {
        private const val ASSET_BASE_FOLDER = "words"
    }

}